File: unit tests/cpp/simple_unit_tests.cpp
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

// Clase Coordenada
class Coordenada
{
public:
    float x, y, z;

    Coordenada(float x, float y, float z) : x(x), y(y), z(z) {}
};

// Clase SistemaLuminoso
class SistemaLuminoso
{
public:
    bool estado;
    int frecuencia;
    char color;

    SistemaLuminoso(bool estado, int frecuencia, char color)
        : estado(estado), frecuencia(frecuencia), color(color)
    {
        if (color != 'R' && color != 'G' && color != 'B')
        {
            throw invalid_argument("Color invalido: debe ser 'R', 'G' o 'B'");
        }
    }
};

// Clase PuntoDeTrayectoria
class PuntoDeTrayectoria
{
public:
    Coordenada coordenada;
    int tiempo;
    SistemaLuminoso sistemaLuminoso;

    PuntoDeTrayectoria(Coordenada coord, int tiempo, SistemaLuminoso sistema)
        : coordenada(coord), tiempo(tiempo), sistemaLuminoso(sistema) {}
};

// Clase Trayectoria que agrega objetos de tipo PuntoDeTrayectoria
class Trayectoria
{
private:
    vector<PuntoDeTrayectoria> puntos;
    const int MAX_PUNTOS = 10;

public:
    bool agregarPunto(const PuntoDeTrayectoria &punto)
    {
        if (puntos.size() < MAX_PUNTOS)
        {
            puntos.push_back(punto);
            return true;
        }
        return false;
    }

    bool eliminarUltimoPunto()
    {
        if (!puntos.empty())
        {
            puntos.pop_back();
            return true;
        }
        return false;
    }

    int obtenerCantidadPuntos() const
    {
        return puntos.size();
    }
};

// Pruebas unitarias
void pruebasUnitarias()
{
    Trayectoria trayectoria;

    Coordenada c1(10.5, 20.3, 30.1);
    SistemaLuminoso s1(true, 50, 'R');
    PuntoDeTrayectoria p1(c1, 100, s1);

    assert(p1.sistemaLuminoso.estado == true);
    assert(p1.sistemaLuminoso.frecuencia == 50);
    assert(p1.sistemaLuminoso.color == 'R');

    assert(trayectoria.agregarPunto(p1) == true);
    assert(trayectoria.obtenerCantidadPuntos() == 1);
    assert(trayectoria.eliminarUltimoPunto() == true);  // Eliminar el punto agregado
    assert(trayectoria.obtenerCantidadPuntos() == 0);   // Verificar cantidad
    assert(trayectoria.eliminarUltimoPunto() == false); // No hay puntos para eliminar

    Coordenada c2(15.0, 25.4, 35.2);
    SistemaLuminoso s2(false, 60, 'G');
    PuntoDeTrayectoria p2(c2, 150, s2);

    // Intento de agregar puntos más allá del límite
    for (int i = 0; i < 10; ++i)
    {
        cout << trayectoria.obtenerCantidadPuntos() << endl;
        assert(trayectoria.agregarPunto(p2) == true);
    }
    assert(trayectoria.agregarPunto(p2) == false);

    try
    {
        SistemaLuminoso s_invalido(true, 50, 'Z');
        assert(false);
    }
    catch (const invalid_argument &e)
    {
        assert(true);
    }

    // Verificar que la cantidad de puntos no exceda el límite
    assert(trayectoria.obtenerCantidadPuntos() <= 10);
}

int main()
{
    pruebasUnitarias();
    cout << "Todas las pruebas pasaron exitosamente!" << endl;

    return 0;
}

File: unit tests/python/simple_unit_tests.py
class Coordenada:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

class SistemaLuminoso:
    def __init__(self, estado, frecuencia, color):
        self.estado = estado
        self.frecuencia = frecuencia
        self.color = color

class PuntoDeTrayectoria:
    def __init__(self, coordenada, tiempo, sistema_luminoso):
        self.coordenada = coordenada
        self.tiempo = tiempo
        self.sistema_luminoso = sistema_luminoso

class Trayectoria:
    def __init__(self):
        self.puntos = []
        self.MAX_PUNTOS = 10

    def agregar_punto(self, punto):
        if len(self.puntos) < self.MAX_PUNTOS:
            self.puntos.append(punto)
            return True
        return False

    def mostrar_datos(self):
        print("X\tY\tZ\tTiempo\tEstado\tFrecuencia\tColor")
        for punto in self.puntos:
            estado = "Activo" if punto.sistema_luminoso.estado else "Inactivo"
            print(f"{punto.coordenada.x}\t{punto.coordenada.y}\t{punto.coordenada.z}\t"
                  f"{punto.tiempo}\t{estado}\t{punto.sistema_luminoso.frecuencia}\t"
                  f"{punto.sistema_luminoso.color}")

    def obtener_cantidad_puntos(self):
        return len(self.puntos)

# Pruebas unitarias
def pruebas_unitarias():
    trayectoria = Trayectoria()

    c1 = Coordenada(10.5, 20.3, 30.1)
    s1 = SistemaLuminoso(True, 50, 'R')
    p1 = PuntoDeTrayectoria(c1, 100, s1)

    assert trayectoria.agregar_punto(p1) is True
    assert trayectoria.obtener_cantidad_puntos() == 1

    c2 = Coordenada(15.0, 25.4, 35.2)
    s2 = SistemaLuminoso(False, 60, 'G')
    p2 = PuntoDeTrayectoria(c2, 150, s2)

    assert trayectoria.agregar_punto(p2) is True
    assert trayectoria.obtener_cantidad_puntos() == 2

    # Intento de agregar más puntos hasta superar el límite
    for _ in range(8):
        assert trayectoria.agregar_punto(p2) is True
    assert trayectoria.agregar_punto(p2) is False  # Debería fallar al superar el límite

if __name__ == "__main__":
    pruebas_unitarias()
    print("Todas las pruebas pasaron exitosamente!")
